src/java.base/share/classes/sun/security/pkcs/SignerInfo.java

```diff
@@ -471,6 +501,57 @@ SignerInfo verify(PKCS7 block, byte[] data)
         return null;
     }
 
+    /**
+     * Derives the signature algorithm name from the digest algorithm
+     * name and the encryption algorithm name inside a PKCS7 SignerInfo.
+     *
+     * For old style PKCS7 files where we use RSA, DSA, EC asencAlgId
+     * a DIGESTwithENC algorithm is returned. For new style RSASSA-PSS
+     * and EdDSA encryption, this method ensures digAlgId is compatible
+     * with the algorithm.
+     *
+     * @param digAlgId the digest algorithm
+     * @param encAlgId the encryption or signature algorithm
+     * @param directSign whether the signature is calculated on the content
+     *                   directly. This makes difference for Ed448.
+     */
+    public static String makeSigAlg(AlgorithmId digAlgId, AlgorithmId encAlgId,
+            boolean directSign) throws NoSuchAlgorithmException {
+        String encAlg = encAlgId.getName();
+        if (encAlg.contains("with")) {
+            return encAlg;
+        }
+        switch (encAlg) {
+            case "RSASSA-PSS":
+                PSSParameterSpec spec = (PSSParameterSpec)
+                        SignatureUtil.getParamSpec(encAlg, encAlgId.getParameters());
+                if (!AlgorithmId.get(spec.getDigestAlgorithm()).equals(digAlgId)) {
+                    throw new NoSuchAlgorithmException("Incompatible digest algorithm");
+                }
+                return encAlg;
+            case "Ed25519":
+                if (!digAlgId.equals(SignatureUtil.EdDSADigestAlgHolder.sha512)) {
+                    throw new NoSuchAlgorithmException("Incompatible digest algorithm");
+                }
+                return encAlg;
+            case "Ed448":
+                if (directSign) {
+                    if (!digAlgId.equals(SignatureUtil.EdDSADigestAlgHolder.shake256)) {
+                        throw new NoSuchAlgorithmException("Incompatible digest algorithm");
+                    }
+                } else {
+                    if (!digAlgId.equals(SignatureUtil.EdDSADigestAlgHolder.shake256$512)) {
+                        throw new NoSuchAlgorithmException("Incompatible digest algorithm");
+                    }
+                }
+                return encAlg;
+            default:
+                String digAlg = digAlgId.getName().replace("-", "");
```

Comment:
Good suggestion. I'll also try some tests.

Reply:
In fact, since now I directly write the signature algorithm into the `SignerInfo.digestEncryptionAlgorithmId` field, the code above is not used at all. The `makeSigAlg` method directly returns the `encAlgId` argument if it has "with" inside.

I'll fix it anyway.